### PR TITLE
[tensorexpr] Memory planning: add pre-allocation for reused intermediate buffers

### PR DESCRIPTION
**Pull request by huiguoo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66967

Differential Revision: [D31810243](https://our.internmc.facebook.com/intern/diff/D31810243)

### CODE REVIEW

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Review by t10-13rocket — APPROVED** [review]
LGTM

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

**Comment by ZolotukhinM** [comment]
Do you mind elaborate on what this PR is doing (ideally, in the commit message)? Currently it's somewhat unclear: we've already had pre-allocation and reusing for intermediate buffers, what are we adding/changing here and why?

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

File: torch/csrc/jit/tensorexpr/codegen.h [hunk 1]
```diff
@@ -111,6 +119,7 @@ class TORCH_API CodeGen {
   std::vector<BufferArg> buffer_args_;
   at::Device device_ = at::kCPU;
   std::string kernel_func_name_ = "func";
+  std::vector<BufPtr> buffers_to_alloc_;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/codegen.h:122]: The names are getting quite confusing now. If one reads this code after this PR lands, they would see `buffer_to_alloc`, `buf_allocs`, `interm_bufs`, `buf_args` and many of those things, despite appearing similar, mean very different things. Let's rename them to better reflect what they hold. E.g. one of them can be `preallocated_bufs` - i.e. a set of list of buffers that don't have Allocate/Free statements and need to be preallocated outside of the codegen (i.e. in the Kernel). The other one can be `dynamically_allocated_bufs` - those are the intermediate bufs that have Allocate/Free/Alias stmt for them. You could choose different names if you'd like but please try to assess the code from a position of a person reading it for the first time.

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp [hunk 2]
```diff
@@ -332,13 +332,14 @@ LLVMCodeGen::LLVMCodeGen(
...
+    const bool pre_alloc,
     Dtype dtype,
     c10::optional<std::string> triple,
     c10::optional<std::string> cpu,
     c10::optional<std::string> attrs)
-    : CodeGen(stmt, args, device, kernel_func_name) {
+    : CodeGen(stmt, args, device, kernel_func_name, pre_alloc) {
   impl_ = std::make_unique<LLVMCodeGenImpl>(
-      this->stmt(), args, device, dtype, triple, cpu, attrs);
+      this->stmt(), this->buffer_args(), device, dtype, triple, cpu, attrs);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/llvm_codegen.cpp:342]: Why is this change needed?

File: torch/csrc/jit/tensorexpr/codegen.cpp [hunk 3]
```diff
@@ -185,34 +186,45 @@ std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
...
+    // Only allocate buffers that are mapped to selves.
     if (rit->first == rit->second) {
       BufPtr buf = rit->first;
-      b->prepend_stmt(alloc<Allocate>(buf));
-      b->append_stmt(alloc<Free>(buf));
+      if (pre_alloc) {
+        // Save buffers for buffer preallocation in TEK.
+        buffers_to_pre_alloc_.emplace_back(buf);
+        // Add preallocated buffers to args.
+        buffer_args_.emplace_back(BufHandle(buf));
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/codegen.cpp:208]: It seems like if `pre_alloc` is true, but the buffer has dynamic shapes, it won't be allocated at all. Here we would expect TEK to pre-allocate it, but TEK would reject to do that since shapes are not constants. Am I missing something? Do you mind adding a test to exercise this?

File: torch/csrc/jit/tensorexpr/codegen.cpp [hunk 4]
```diff
@@ -185,34 +186,45 @@ std::vector<std::pair<BufPtr, BufPtr>> AllocBufsWithMemReuse(
...
+  set_stmt(b);
 }
 
 // We allocate intermediate buffers by inserting Allocate/Free or
 // PlacementAllocate stmts. Allocate/Free stmts will allocate memory at runtime,
 // and PlacementAllocate stmt reuses the memory of one buffer for another
 // buffer. In current implementation, we use linear scan for memory reuses.
 // TODO: try more memory reuse algorithms and compare their memory efficiency.
-void CodeGen::allocIntermediateBufs() {
+void CodeGen::allocIntermediateBufs(const bool pre_alloc) {
```
[the remark this answers is not in the document]
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/codegen.cpp:227]: Yeah. We can keep it for better coding though. The future reviewers would know it's read only.

File: torch/csrc/jit/tensorexpr/llvm_codegen.cpp (hunk 2 above)
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/llvm_codegen.cpp:342]: Yes it's needed. The buffer args will be updated in the constructor of its parent (CodeGen).

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31810243).

File: torch/csrc/jit/tensorexpr/codegen.cpp (hunk 3 above)
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/codegen.cpp]: That's a good point! Have fixed it and added a test.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 5]
```diff
@@ -1335,17 +1328,18 @@ std::vector<BufPtr> TensorExprKernel::preAllocIntermediateBufs(
...
-      continue;
+      throw std::runtime_error(
+          "TensorExprKernel: preallocation does not support dynamic shapes!");
     }
     auto bp = (void*)malloc(size);
     if (!bp) {
-      remaining_interm_bufs.push_back(buf);
-      continue;
+      throw std::runtime_error(
+          "TensorExprKernel: memory allocation failed during preallocation!");
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1337]: Let's use one of TE exceptions instead of generic runtime_error here (see https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/exceptions.h).

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 6]
```diff
@@ -1585,7 +1579,13 @@ void TensorExprKernel::compile() {
       stmt_,
       bufferArgs_,
       device_,
-      kernel_func_name_);
+      kernel_func_name_,
+      pre_alloc_);
+
+  if (pre_alloc_) {
```
[the remark this answers is not in the document]
Comment (huiguoo) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1585]: I understand. This is what we need to go through to build good things. Thanks for letting me know. Otherwise it would take more time and effort for me to maintain this PR.

**Comment by huiguoo** [comment]
Most of the changes in this PR are not useful anymore. Closing the PR.
